ir/src/constraints/mod.rs

```diff
@@ -72,10 +73,11 @@ impl ExprDetails {
 /// for each constraint against the main trace.
 #[derive(Default, Debug)]
 pub(super) struct Constraints {
-    /// Constraint roots for all boundary constraints against the execution trace, by trace segment,
-    /// where boundary constraints are any constraints that apply to either the first or the last
-    /// row of the trace.
-    boundary_constraints: Vec<Vec<ConstraintRoot>>,
+    /// Constraint roots for all boundary constraints against the execution trace, grouped by trace
+    /// segment. They map the constrained boundary within the trace to the [ConstraintRoot] that
+    /// contains its [ConstraintDomain] and entry [NodeIndex] in the graph. Boundary constraints are
+    /// any constraints that apply to either the first or the last row of the trace.
+    boundary_constraints: Vec<BTreeMap<ConstrainedBoundary, ConstraintRoot>>,
```

Comment:
The information here seems somewhat redundant - both `ConstrainedBoundary` and `ConstraintRoot` contain domain info. Moreover, `ConstrainedBoundary` also contains `trace_segment` which seems to be the index into top-level vector here.

It seems to me that putting `trace_segment` into `ConstrainedBoundary` may not be necessary? But more generally, it seems to me that the `BTreeMap` and `ConstrainedBoundary` struct are used primarily for error-checking purposes (to make sure we don't have multiple boundary constraints defined against the same boundary of the same column). If so, I'd rather use something else for error checking and keep this as `Vec<Vec<ConstraintRoot>>`. For example, we could introduce another instance variable similar to `variable_roots` but to keep track of which boundary constraints are already "taken".

Separately, looking at the structure below, I wonder if we need to keep two separate member variables for validity and transition constraints. Could we not track them in a single vector as:
```Rust
pub(super) struct Constraints {
    ...
    integrity_constraints: Vec<Vec<ConstraintRoot>>,
    ...
}
```
The information contained in `ConstraintRoot` should be sufficient to differentiate between integrity and transition constraints. Or am I missing something?

Anyways, this is probably a separate issue not related to this PR.

Reply:
> It seems to me that putting trace_segment into ConstrainedBoundary may not be necessary?

Yeah, I originally implemented it this way and then got confused & changed it at the last minute, which I have been thinking was a mistake. I'll remove it from there.

> But more generally, it seems to me that the BTreeMap and ConstrainedBoundary struct are used primarily for error-checking purposes (to make sure we don't have multiple boundary constraints defined against the same boundary of the same column). If so, I'd rather use something else for error checking

I was on the fence about this and considered this approach too. The reason I did it this way was to avoid tracking the same thing twice, but I'm happy to try something else. In fact, if we keep the `ConstrainedBoundary` as it is (with the `trace_segment`) then this could be tracked more cleanly than my original thought here.

> Separately, looking at the structure below, I wonder if we need to keep two separate member variables for validity and transition constraints. Could we not track them in a single vector

Yes, we could track them as a single vector, but it seemed like it gave a bit more flexibility to have them separated if we want to access them separately without having to filter the vectors. I don't feel very strongly, but there's no cost to separating them as we process them, and there's a small cost to doing so later, which is why I had previously done it this way.

Reply:
I've updated this - please let me know what you think of the adjusted approach for tracking constrained boundaries